### PR TITLE
add envoy admin api server state health check to pilot wait

### PR DESCRIPTION
**Pull request by tariq1890**
Please provide a description for what this PR is for.

We tested this feature and found that the Istio Proxy readiness probe check doesn't necessarily indicate that it's ready to accept traffic. We have to wait for the Envoy Proxy server state to go `LIVE`. This PR adds that check as an extra step

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[X] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/cmd/pilot-agent/wait.go [hunk 1]
```diff
@@ -57,6 +61,17 @@ var (
 )
 
 func checkIfReady(client *http.Client, url string) error {
+	// Check if the Envoy Proxy Server is in LIVE state
+	info, err := envoy.GetServerInfo(envoyAdminPort)
```
Comment (ramaraochavali) [inline pilot/cmd/pilot-agent/wait.go:65]: Can you use this https://github.com/istio/istio/blob/master/pilot/cmd/pilot-agent/status/util/stats.go#L65 - that what is used to detect that Envoy is ready

File: pilot/cmd/pilot-agent/wait.go [hunk 2]
```diff
@@ -57,6 +61,17 @@ var (
...
+	if err != nil {
+		return err
+	}
+	if info.State != envoyAdmin.ServerInfo_LIVE {
+		// We're live!
```
Comment (shamsher31) [inline pilot/cmd/pilot-agent/wait.go:70]: I think no need for this comment.

File: pilot/cmd/pilot-agent/wait.go (hunk 1 above)
Comment (howardjohn) [reply pilot/cmd/pilot-agent/wait.go:65]: Why do we need any of this code anyways? It should be hitting the readiness probe which will in turn check the state for us?

Comment (tariq1890) [reply pilot/cmd/pilot-agent/wait.go:65]: @ramaraochavali the `checkIfReady` code already calls that method. Unfortunately, the readiness stats check does not necessarily indicate traffic readiness. At least from my experience so far.

File: pilot/cmd/pilot-agent/wait.go (hunk 2 above)
Comment (tariq1890) [reply pilot/cmd/pilot-agent/wait.go:70]: Sure

**Comment by tariq1890** [comment]
@howardjohn I can move the Envoy Live check to the readiness probe Handler method; would that be an acceptable change?

**Comment by howardjohn** [comment]
It's already doing it afaik

**Comment by tariq1890** [comment]
So it checks the `stats` endpoint instead of the `server_info`. I don't see why they should be any different.

But FWIW, I changed the lifecycle hook to wait on `localhost:15000/server_info | grep LIVE` instead of `localhost:15021/healthz/ready` and it started working for me.

**Comment by kyessenov** [comment]
I've just hit something similar in https://github.com/envoyproxy/envoy/issues/12245. The only real way to test is to mock a fast integration test since the delay is likely to be O(ms). Envoy is ready to accept if xDS is accepted and workers have started. LIVE is tied to the server init manager at the start-up which is not a good indicator of readiness since the event loop may not have started. It might be best to issue a few test health queries for this to be usable in prod.

**Comment by tariq1890** [comment]
Ideally, this should be a part of the readiness probe endpoint.  From my understanding of the linked threads, the `workers_started` does not necessarily indicate traffic readiness. Perhaps we should have a `workers_ready` field?

**Comment by howardjohn** [comment]
The readiness probe does actually go through a real envoy listener now. But
it's a static one so maybe it behaves differently

**Comment by lambdai** [comment]
@howardjohn  which static listener?

**Comment by howardjohn** [comment]
https://github.com/istio/istio/blob/e63ab2161fde82e84d03e5ab7736d715a5879276/tools/packaging/common/envoy_bootstrap.json#L501

**Comment by tariq1890** [comment]
I am okay with closing this PR and opening an issue. I am a bit confused by most of the responses here though. Are we at least acknowledging that there is an issue here that needs to be fixed?

**Comment by lambdai** [comment]
@howardjohn Thanks!

That static listener should be an issue. The traffic of that health check listener starts to serve the http request along with other dynamic listeners

**Comment by tariq1890** [comment]
Any updates on this?

**Comment by tariq1890** [comment]
Friendly ping for this :)

**Comment by tariq1890** [comment]
@ramaraochavali Thoughts on this?

**Comment by ramaraochavali** [comment]
I do not see any problem here the /server_info and stat both represent the same thing and we do check that here https://github.com/istio/istio/blob/5f78dfca600be273f50c2f053cfb3f6037b38035/pilot/cmd/pilot-agent/status/util/stats.go#L78 before marking proxy ready along with workers_started. 

Regarding the second point about `workers_started` not indicating the server readiness to accept traffic - it was actually added to solve the same problem as I mentioned here https://github.com/envoyproxy/envoy/issues/12245#issuecomment-662856061. I think the documentation needs to be updated to reflect that part as was mentioned there.

As @howardjohn mentioned, the healthness probe goes via Envoy listener indicating that it is actually serving traffic. Till workers are started, even if the listener is created - it wont serve traffic. 


If you are able to reproduce the problem consistently - can you please share Envoy logs + `/stats?usedonly&filter=^(server\\.state|listener_manager\\.workers_started)` output?

**Comment by tariq1890** [comment]
@ramaraochavali I am unsure how to reproduce this. 

I can explain my use case, however:

i) We are running helm tests for our Chart deployments. Since this helm test makes outbound https calls, it should only be deployed after the istio-proxy is ready. If not, the helm test will fail (because of the failed outbound https call) and thus fail the whole chart deployment.

ii) We've already enabled `.holdApplicationUntilProxyStarts`, so the `istio-proxy` is pushed to the front of the container list. But from our observations, we find that the istio-proxy readiness is determined pre-maturely as main container of helm test pod still fails when it is run after starting the istio-proxy container.

iii) To work around this issue, we added a custom lifecycle hook and this has solved our problem. Our helm tests and applications (their network traffic) work consistently when injected with the istio-proxy. This is our custom lifecycle hook:

```

    lifecycle:
      postStart:
        exec:
          command: ["/bin/sh", "-c", "while ! curl -s localhost:15000/server_info | grep LIVE ; do sleep 5; done"]
```

**Comment by tariq1890** [comment]
As per @kyessenov , this may not be the right approach, but it seems to work rather well.

If this is an unreliable check, we are willing to move to a better approach, but the current `checkIfReady` check is simply not working for us

**Comment by ramaraochavali** [comment]
@tariq1890 Can u please proxy logs from you tests if possible?

**Pull request closed without merging**